### ISSUE TITLE
Proposal: More parallelism in xUnit

### ISSUE THREAD

**Issue by SetTrend** [xunit/xunit#595]
Currently, to have a number of tests run in parallel, you need to create separate classes for each of them:

``````c#
public class TestClass1
{
    [Fact]
    public void Test1()
    {
        Thread.Sleep(3000);
    }
}

public class TestClass2
{
    [Fact]
    public void Test2()
    {
        Thread.Sleep(5000);
    }
}
``````

This is a particular nuisance on unit tests which do not cause any side effects, like tests just testing for invalid arguments.

I would therefore very much suggest all tests of a single class to run in parallel by default unless the class has a construtor defined, or an additional, optional string parameter to the `FactAttribute` (e.g. *CollectionName*) enforces the creation of inter-class collections to override the parallelism if no constructor has been defined:

``````c#
public class NonParallelTestClass
{
    public NonParallelTestClass() {}    // constructor indicates test preparation to occur, hence no parallelism by default.


    [Fact]
    public void Test1()    // runs immediately
    {
        Thread.Sleep(3000);
    }

    [Fact]
    public void Test2()    // runs after Test1() has completed
    {
        Thread.Sleep(5000);
    }

    [Fact(CollectionName="Separate collection")]
    public void Test3()    // runs immediately
    {
        Thread.Sleep(2000);
    }

    [Fact(CollectionName="Separate collection")]
    public void Test4()    // runs after Test3() has completed
    {
        Thread.Sleep(3000);
    }

    [Fact(CollectionName="Another collection")]
    public void Test5()    // runs immediately
    {
        Thread.Sleep(9000);
    }
}



public class ParallelTestClass    // no constructor defined, hence paralleism occurs by default.
{
    [Fact]
    public void Test1()    // runs immediately
    {
        Thread.Sleep(3000);
    }

    [Fact]
    public void Test2()    // runs immediately
    {
        Thread.Sleep(5000);
    }

    [Fact(CollectionName="Keep apart")]
    public void Test3()    // runs immediately
    {
        Thread.Sleep(2000);
    }

    [Fact(CollectionName="Keep apart")]
    public void Test4()    // runs after Test3() has completed
    {
        Thread.Sleep(3000);
    }
}
``````

**Comment by bradwilson** [comment]
Related: https://github.com/xunit/xunit/issues/286

TL;DR: We don't intend to add this to core xUnit.net, but it would be possible to implement this feature using a custom test framework (which lets you completely control how things are discovered and run).
